### ISSUE TITLE
Adding support for additional SSH enabled devices

### ISSUE THREAD

**Issue by username-is-already-taken2** [inspec/inspec#448]
Hello there

I would like to be able to run compliance checks natively against my ssh enabled firewall (Juniper Netscreen) but I'm having a problem getting inspec to open a shell on the device. (I can ssh to the device using putty fine)

![inspec-error1](https://cloud.githubusercontent.com/assets/13700297/12976586/22363afc-d0bd-11e5-855d-e6f46a7da990.PNG)

![inspec-error2](https://cloud.githubusercontent.com/assets/13700297/12976588/285d6cb6-d0bd-11e5-9084-4cbd8a8ec39b.PNG)

other then use the --diagnose command is there any other debug I might be able to run to assist?  if I can get it working then I will be adding over 100nodes to my server.

Best Regards

Gary

**Comment by srenatus** [comment]
Hi Gary,

Thanks for the report and your offer of assistance. Talking to _any_ ssh-enabled device is certainly the goal for inspec, or rather train.

As you noticed, the `--diagnose` command is of little use beyond helping with the question _"where does this configuration value come from"_.

On the internet, people seem to be able to talk successfully with JunOS devices and Net::SSH. [This project](), however, [is using a PTY for that](https://github.com/ytti/seeker-junos/blob/master/lib/seeker/ssh.rb#L36). Missing support for working with ssh servers that require a PTY has already been [found as an issue in train, tracked here](https://github.com/chef/train/issues/60). Note that _before your bug report_, we had the consensus that
1. a PTY is usually only required for `sudo`, and
2. if the system is configured to _not allow non-interactive sudo session_ and _we want to do just that_, working around the problem isn't advisable and we should rather fix the system (and make _fixing the system_ easy)

Now if Jupiter devices can only be accessed with a PTY, this assessment has to be re-evaluated.

I haven't got access to any Juniper device, but here's something I'd ask you to try: connect to your networking equipment _without_ a PTY. This is possible with the OpenSSH client:

```
ssh user@device SOME_COMMAND
```

where `SOME_COMMAND` is anything the router understands, e.g. `show interfaces fe-1/3/1 statistics`.
If this _doesn't work_, but 

```
ssh -t user@device SOME_COMMAND
```

or

```
ssh -tt user@device SOME_COMMAND
```

does, then it's most likely a PTY issue. You can get an openssh-client for Windows using CYGWIN or MSYS2. It'd be easiest if you had a *nix box at your disposal, though.

Note that while there are [special libraries for Rubying with Juniper's](https://github.com/Juniper/net-netconf), this would pretty much defeat the purpose of "let's just talk with anything that talks standard protocols" :confused:

Edit: I'm not implying that having a Juniper transport plugin in train would be a bad thing, though. It's just a matter of how far we can stretch with ssh-only ;)

Thanks a lot for your assistance,
Stephan

**Comment by username-is-already-taken2** [comment]
Thanks Stephan for your reply

I tried what you suggested using a openssh on an ubuntu vm, here are the results. The last two commands are a repeat of the one with and without the -tt but with the -verbose flag set.

**`eeyore@ubuntu:~$ dpkg -s openssh-client | grep Version:`**
`Version: 1:6.6p1-2ubuntu2.6`

**`eeyore@ubuntu:~$ ssh netscreen@x.x.x.x "get system version"`**
`netscreen@x.x.x.x's password:`
`eeyore@ubuntu:~$`

**`eeyore@ubuntu:~$ ssh -t netscreen@x.x.x.x "get system version"`**
`netscreen@x.x.x.x's password:`
`Connection to x.x.x.x closed.`

**`eeyore@ubuntu:~$ ssh -tt netscreen@x.x.x.x "get system version"`**
`netscreen@x.x.x.x's password:`
`Connection to x.x.x.x closed by remote host.`
`Connection to x.x.x.x closed.`
`eeyore@ubuntu:~$`

**`eeyore@ubuntu:~$ ssh -tt -v netscreen@x.x.x.x "get system version"`**
`OpenSSH_6.6.1, OpenSSL 1.0.1f 6 Jan 2014`
`debug1: Reading configuration data /etc/ssh/ssh_config`
`debug1: /etc/ssh/ssh_config line 19: Applying options for *`
`debug1: Connecting to x.x.x.x [x.x.x.x] port 22.`
`debug1: Connection established.`
`debug1: identity file /home/eeyore/.ssh/id_rsa type -1`
`debug1: identity file /home/eeyore/.ssh/id_rsa-cert type -1`
`debug1: identity file /home/eeyore/.ssh/id_dsa type -1`
`debug1: identity file /home/eeyore/.ssh/id_dsa-cert type -1`
`debug1: identity file /home/eeyore/.ssh/id_ecdsa type -1`
`debug1: identity file /home/eeyore/.ssh/id_ecdsa-cert type -1`
`debug1: identity file /home/eeyore/.ssh/id_ed25519 type -1`
`debug1: identity file /home/eeyore/.ssh/id_ed25519-cert type -1`
`debug1: Enabling compatibility mode for protocol 2.0`
`debug1: Local version string SSH-2.0-OpenSSH_6.6.1p1 Ubuntu-2ubuntu2.6`
`debug1: Remote protocol version 2.0, remote software version NetScreen`
`debug1: no match: NetScreen`
`debug1: SSH2_MSG_KEXINIT sent`
`debug1: SSH2_MSG_KEXINIT received`
`debug1: kex: server->client 3des-cbc hmac-sha1 none`
`debug1: kex: client->server 3des-cbc hmac-sha1 none`
`debug1: sending SSH2_MSG_KEXDH_INIT`
`debug1: expecting SSH2_MSG_KEXDH_REPLY`
`debug1: Server host key: DSA 5e:9e:42:db:7a:d0:66:bd:57:a5:98:08:a0:30:95:e4`
`debug1: Host 'x.x.x.x' is known and matches the DSA host key.`
`debug1: Found key in /home/eeyore/.ssh/known_hosts:1`
`debug1: ssh_dss_verify: signature correct`
`debug1: SSH2_MSG_NEWKEYS sent`
`debug1: expecting SSH2_MSG_NEWKEYS`
`debug1: SSH2_MSG_NEWKEYS received`
`debug1: SSH2_MSG_SERVICE_REQUEST sent`
`debug1: SSH2_MSG_SERVICE_ACCEPT received`
`debug1: Authentications that can continue: password`
`debug1: Next authentication method: password`
`netscreen@x.x.x.x's password:`
`debug1: Authentication succeeded (password).`
`Authenticated to x.x.x.x ([x.x.x.x]:22).`
`debug1: channel 0: new [client-session]`
`debug1: Entering interactive session.`
`debug1: Sending environment.`
`debug1: Sending env LANG = en_US.UTF-8`
`debug1: Sending command: get system version`
`debug1: client_input_channel_req: channel 0 rtype exit-status reply 0`
`debug1: channel 0: free: client-session, nchannels 1`
`Connection to x.x.x.x closed.`
`Transferred: sent 2856, received 1000 bytes, in 0.2 seconds`
`Bytes per second: sent 17972.7, received 6293.0`
`debug1: Exit status 1`
`eeyore@ubuntu:~$` 

**`eeyore@ubuntu:~$ ssh -v netscreen@172.31.4.67 "get system version"`**
`...`
`netscreen@172.31.4.67's password:`
`debug1: Authentication succeeded (password).`
`Authenticated to 172.31.4.67 ([172.31.4.67]:22).`
`debug1: channel 0: new [client-session]`
`debug1: Entering interactive session.`
`debug1: Sending environment.`
`debug1: Sending env LANG = en_US.UTF-8`
`debug1: Sending command: get system version`
`debug1: client_input_channel_req: channel 0 rtype exit-status reply 0`
`debug1: channel 0: free: client-session, nchannels 1`
`Connection to 172.31.4.67 closed by remote host.`
`Transferred: sent 2544, received 984 bytes, in 2.5 seconds`
`Bytes per second: sent 1017.4, received 393.5`
`debug1: Exit status 1`
`eeyore@ubuntu`

As you see I wasn't very successful :( I don't know where that leaves us.

I do have a few other devices I was planning on creating compliance rules for, I might work my way through those today and see which ones I can run commands against.

**Comment by srenatus** [comment]
Hey Gary,

thanks for your input, it's very comprehensive. :+1:

And it is quite interesting, and not what I was expecting: so irregardless of whether you request a PTY or not, _it fails_. It's correct that _success_ would have meant some output in one of the three cases?

One item I forgot on my wishlist -- could you try `ssh -T netscreen@172.31.4.67 "get system version"`? This will explicitly _not_ request a PTY. (I am a little confused about what ssh client's default really does...let's just be complete.)

It's quite hard to find information about ScreenOS. Apparently, people do have problems using `openssh`, and [one workaround is setting ControlMaster:](http://kb.juniper.net/InfoCenter/index?page=content&id=KB12409&actp=search)

```
ssh -oControlMaster=auto hostname
```

Looking at that KB entry, I'll venture another stab in the dark: [PuTTY supports SSHv1 by default](http://blather.michaelwlucas.com/archives/886), and openssh uses SSHv2 (note the `SSH2_MSG_*` commands). **Could you verify that your NetScreen device has SSHv2 enabled?** You could try connecting with PuTTY and SSHv1 _disabled_ or use the Juniper CLI to figure it out.

Disabling SSHv1 in PuTTY (from the blog post linked above): _Open PuTTY. On the left side, go to Connection->SSH. Select “2 only.”_

On a related note, from what you're planning to do with inspec and the network devices, would you appreciate or prefer a more NETCONF-y interface over what the NetScreen CLI commands would give you?

Thanks again
Stephan

**Comment by username-is-already-taken2** [comment]
Hello Again

I tried the -T parameter, but it ended up like the others :(

`eeyore@ubuntu:~$ ssh -T -v netscreen@x.x.x.x "get system version"`
`...`
`netscreen@x.x.x.x's password:`
`debug1: Authentication succeeded (password).`
`Authenticated to x.x.x.x ([x.x.x.x]:22).`
`debug1: channel 0: new [client-session]`
`debug1: Entering interactive session.`
`debug1: Sending environment.`
`debug1: Sending env LANG = en_US.UTF-8`
`debug1: Sending command: get system version`
`debug1: client_input_channel_req: channel 0 rtype exit-status reply 0`
`debug1: channel 0: free: client-session, nchannels 1`
`Transferred: sent 2544, received 984 bytes, in 0.3 seconds`
`Bytes per second: sent 8482.2, received 3280.9`
`debug1: Exit status 1`

Here is the output of the -oControlMaster=auto switch

`eeyore@ubuntu:~$ ssh -oControlMaster=auto netscreen@x.x.x.x "get system version"`
`netscreen@x.x.x.x's password:`
`eeyore@ubuntu:~$ ssh -oControlMaster=auto -t netscreen@x.x.x.x "get system version"`
`netscreen@x.x.x.x's password:` 
`Connection to x.x.x.x closed.`
`eeyore@ubuntu:~$ ssh -oControlMaster=auto -tt netscreen@x.x.x.x "get system version"`
`netscreen@x.x.x.x's password:` 
`Connection to x.x.x.x closed.`
`eeyore@ubuntu:~$ ssh -oControlMaster=auto -T netscreen@x.x.x.x "get system version"`
`netscreen@x.x.x.x's password:` 
`eeyore@ubuntu:~$` 

Your correct about the the default to SSH v1 (I am planning on have a compliance check to make sure its enabled :) so I know the config setting.

`eeyore@ubuntu:~$ ssh -1 netscreen@x.x.x.x`
`Protocol major versions differ: 1 vs. 2`
`eeyore@ubuntu:~$ ssh -2 netscreen@x.x.x.x`
`netscreen@x.x.x.x's password:` 
`This is....`
`Firewall(M)->` 

NETCONF could be a winner for when I get round to writing compliance checks Juniper Junos OS Firewalls I have a load of those too :) I'll find out which of my other devices have NETCONF support.

Thanks again for your help

**Comment by srenatus** [comment]
Hey Gary,

So, examining your results (please correct me if I am misled), I come to the conclusion that with your netscreen devices, non-interactive ssh fails, regardless of the use of `-t`, `-tt`, `-T` (i.e. pty-related things). Interactive use of openssh-client works. (We haven't verified that `ssh netscreen@xx.x.x.` actually works, but since `-2` works and should be the default, I'd conclude it works.)

Is that correct? I wonder what's the reason for that... It's quite unfortunate that Juniper doesn't seem to provide virtual machine images of ScreenOS... :confused:

**Comment by username-is-already-taken2** [comment]
Hello Again.

You are correct with your summary, if it will help I can stand-up a netscreen for you to connect to and test with?  Ideally even though it will be a default device, if I could protect it with an IP filter that would be cool.  Let me know if this will help, I could DM the details via twitter.

I've worked my way through some other devices within my enviroment to test connectivity 

![device-matrix](https://cloud.githubusercontent.com/assets/13700297/13110910/a752f920-d57a-11e5-8106-8a2ef333b2dd.PNG)

*1 = Connected but wouldn't exec commands
*2 = Echos command but doesn't apply a Carriage return
*3 = ssh_connection.rb:107:in rescue in run_command
*4 = ssh_connection.rb:182:in rescue in establish_connection

One thing that I didn't realize before starting this is that not all ssh implimentations are the same :)

So I got some wins with Fortigate and SRX that I can work with for the time being.  I think on the priority looking at the list behind the netscreen would be esx and asa support.

I hope that helps

**Comment by alexpop** [comment]
This reminds me of the [Expect](http://expect.sourceforge.net/) based automation I've done with network devices at Opsware.

**Comment by srenatus** [comment]
Thank you so much, Gary! :+1: 

This table is very helpful. For one thing, it shows us that PTY's are not an issue when it comes to talking to (this sample of) network devices -- the last four columns are always equal.

And it gives us a great todo list when comparing the **InSpec Shell** column with the one to it's right. The ones where `openssh` works but `inspec` doesn't might be low-hanging fruits.

Thanks a lot for your offer wrt. giving us access to this. I'll surely come back at you here :smile:

It's also helpful that you've included your priorities :+1: 

@alexpop: spot-on, apparently people are using something expect-ish already ([Net::SSH::Telnet](http://net-ssh-telnet.rubyforge.org/)): [SO post](http://stackoverflow.com/questions/10327141/using-ruby-gem-net-ssh-to-configure-a-juniper-router)

**Comment by chris-rock** [comment]
@username-is-already-taken2 It may worth to reevaluating the list of devices. We improved train to accept more ssh methods to transmit passwords. Would love to see the updated results.

**Comment by chris-rock** [comment]
https://github.com/chef/train/pull/247 fixed the detection for Cisco devices via ssh

**Comment by arlimus** [comment]
@username-is-already-taken2 Hi Gary, i hope these last fixes are bringing some of the Cisco equipment closer. Let's get in touch if you see something I might have missed. Also there is still some stability work to be done on ssh in cisco IOS12

**Comment by kekaichinose** [comment]
__What’s happening? Why was this issue closed?__
This issue was closed due to some much needed review of legacy issues or issues that were spawned in older versions of InSpec, i.e. < v3.

__Why do I care?__
You would care about this if this was an issue that you are still seeing and/or feel needs to be addressed in the current version of InSpec.

__What do I need to do?__
If this issue is no longer important, no further action is necessary. However, if you think this is something that should be addressed, please open a new issue and refer to the original issue in the description.
